### PR TITLE
opstate - add operational state test scripts

### PR DESCRIPTION
**Pull request by duyen-tran**
Added python scripts for operational state:
TC_OPSTATE_2_1.PY
TC_OPSTATE_2_3.PY

### CODE REVIEW

File: src/python_testing/TC_OPSTATE_2_1.py [hunk 1]
```diff
@@ -125,26 +125,22 @@ async def test_TC_OPSTATE_2_1(self):
...
-                if self.matter_test_config.global_test_params['PIXIT_OPSTATE_ST_PAUSED'] == 1:
-                    self.print_step("6e", "Manually put the device in the paused state")
-                    input("Press Enter when done.\n")
-                    await self.read_and_validate_opstate(step="6f", expected_state=Clusters.OperationalState.Enums.OperationalStateEnum.kPaused)
-            if 'PIXIT_OPSTATE_ST_ERROR' in self.matter_test_config.global_test_params:
-                if self.matter_test_config.global_test_params['PIXIT_OPSTATE_ST_ERROR'] == 1:
-                    self.print_step("6g", "Manually put the device in the error state")
-                    input("Press Enter when done.\n")
-                    await self.read_and_validate_opstate(step="6h", expected_state=Clusters.OperationalState.Enums.OperationalStateEnum.kError)
+            if self.check_pics("OPSTATE.M.ST_STOPPED"):
```
Comment (plauric) [inline src/python_testing/TC_OPSTATE_2_1.py:128]:
```suggestion
            if self.check_pics("OPSTATE.S.M.ST_STOPPED"):
```

We need to use the "S" (server) name particle too.

File: src/python_testing/TC_OPSTATE_2_1.py [hunk 2]
```diff
@@ -0,0 +1,181 @@
...
+                                or in_range, "OperationalError has an invalid ID value!")
+            if in_range:
+                asserts.assert_true(operational_error.errorStateLabel is not None, "ErrorStateLabel should be populated")
+
+            if self.check_pics("OPSTATE.S.M.ERR_NO_ERROR"):
+                self.print_step("7a", "Manually put the device in the no error state")
+                input("Press Enter when done.\n")
+                await self.read_and_validate_operror(step="7b", expected_error=Clusters.OperationalState.Enums.ErrorStateEnum.kNoError)
+            if self.check_pics("OPSTATE.S.M.ERR_UNABLE_TO_START_OR_RESUME"):
+                self.print_step("7c", "Manually put the device in the unable to start or resume state")
```
Comment (plauric) [inline src/python_testing/TC_OPSTATE_2_1.py:167]:
```suggestion
                self.print_step("7c", "Manually put the device in the unable to start or resume error state")
```

File: src/python_testing/TC_OPSTATE_2_1.py [hunk 3]
```diff
@@ -0,0 +1,181 @@
...
+                input("Press Enter when done.\n")
+                await self.read_and_validate_operror(step="7d", expected_error=Clusters.OperationalState.Enums.ErrorStateEnum.kUnableToStartOrResume)
+            if self.check_pics("OPSTATE.S.M.ERR_UNABLE_TO_COMPLETE_OPERATION"):
+                self.print_step("7e", "Manually put the device in the unable to complete operation state")
```
Comment (plauric) [inline src/python_testing/TC_OPSTATE_2_1.py:171]:
```suggestion
                self.print_step("7e", "Manually put the device in the unable to complete operation error state")
```

File: src/python_testing/TC_OPSTATE_2_1.py [hunk 4]
```diff
@@ -0,0 +1,181 @@
...
+                input("Press Enter when done.\n")
+                await self.read_and_validate_operror(step="7f", expected_error=Clusters.OperationalState.Enums.ErrorStateEnum.kUnableToCompleteOperation)
+            if self.check_pics("OPSTATE.S.M.ERR_COMMAND_INVALID_STATE"):
+                self.print_step("7g", "Manually put the device in the command invalid state")
```
Comment (plauric) [inline src/python_testing/TC_OPSTATE_2_1.py:175]:
```suggestion
                self.print_step("7g", "Manually put the device in the command invalid error state")
```

**Review by plauric — APPROVED** [review]

File: src/python_testing/TC_OPSTATE_2_1.py [hunk 5]
```diff
@@ -0,0 +1,181 @@
...
+        if self.check_pics("OPSTATE.S.A0001"):
+            self.print_step(3, "Read CurrentPhase attribute")
+            current_phase = await self.read_mod_attribute_expect_success(endpoint=self.endpoint, attribute=attributes.CurrentPhase)
+            logging.info("CurrentPhase: %s" % (current_phase))
+
+            if phase_list == NullValue:
+                asserts.assert_true(current_phase == NullValue, "CurrentPhase should be null")
+            else:
+                asserts.assert_true(0 <= current_phase and current_phase <= (phase_list_len - 1),
+                                    "CurrentPhase must be between 0 and (phase-list-size - 1)")
```
Comment (andy31415) [inline src/python_testing/TC_OPSTATE_2_1.py:85]: would it make sense to add the current values for logging purposes? If I see this message I don't know what the current_phase is (is it negative? too large?) and I don't know what the phase_list_len is. Also the check could be using `<` and not do the -1.

File: src/python_testing/TC_OPSTATE_2_1.py [hunk 6]
```diff
@@ -0,0 +1,181 @@
...
+
+        if self.check_pics("OPSTATE.S.A0002"):
+            self.print_step(4, "Read CountdownTime attribute")
+            countdown_time = await self.read_mod_attribute_expect_success(endpoint=self.endpoint,
+                                                                          attribute=attributes.CountdownTime)
+
+            logging.info("CountdownTime: %s" % (countdown_time))
+            if countdown_time is not NullValue:
+                asserts.assert_true(countdown_time >= 0 and countdown_time <= 259200, "CountdownTime must be between 0 and 259200")
```
Comment (andy31415) [inline src/python_testing/TC_OPSTATE_2_1.py:94]: similar comment (and probably applies throughout): when an assert fails, I would like to see all relevant values to see obvious invalid things).

File: src/python_testing/TC_OPSTATE_2_3.py [hunk 7]
```diff
@@ -0,0 +1,156 @@
...
+        self.print_step(3, "Read OperationalStateList attribute")
+        op_state_list = await self.read_mod_attribute_expect_success(endpoint=self.endpoint,
+                                                                     attribute=attributes.OperationalStateList)
+
+        logging.info("OperationalStateList: %s" % (op_state_list))
+
+        defined_states = [state.value for state in Clusters.OperationalState.Enums.OperationalStateEnum
+                          if state is not Clusters.OperationalState.Enums.OperationalStateEnum.kUnknownEnumValue]
+
+        state_ids = [s.operationalStateID for s in op_state_list]
```
Comment (andy31415) [inline src/python_testing/TC_OPSTATE_2_3.py:80]: nit: consider using `{}` or `set([...` to create a set for faster checks

File: src/python_testing/TC_OPSTATE_2_3.py [hunk 8]
```diff
@@ -0,0 +1,156 @@
...
+        logging.info("OperationalState: %s" % (operational_state))
+        asserts.assert_equal(operational_state, Clusters.OperationalState.Enums.OperationalStateEnum.kPaused,
+                             "OperationalState ID should be Paused(0x02)")
+
+        self.print_step(6, "Read CountdownTime attribute")
+        initial_countdown_time = await self.read_mod_attribute_expect_success(endpoint=self.endpoint,
+                                                                              attribute=attributes.CountdownTime)
+        logging.info("CountdownTime: %s" % (initial_countdown_time))
+        if initial_countdown_time is not NullValue:
+            in_range = (1 <= initial_countdown_time and initial_countdown_time <= 259200)
```
Comment (andy31415) [inline src/python_testing/TC_OPSTATE_2_3.py:101]: does 

```py
in_range = (1 <= initial_countdown_time <= 259200)
```

work?

**Review by andy31415 — APPROVED** [review]

**Pull request merged**